File: week3.1/duplicate.py
# 217 problem in leetcode
# time complexity O(n)
class Solution:
    def find_duplicate(self,nums):
        # Approach -1
        unique_nums = set()
        for num in nums:
            if num in unique_nums:
                return True
            else:
                unique_nums.add(num)
        return False
        # Approach -2
        # unique_nums = {}
        # for num in nums:
        #     if num in unique_nums:
        #         return True
        #     else:
        #         unique_nums[num] = 1
        # return False
s = Solution()
print(s.find_duplicate([1,2,3,4,2]))
print(s.find_duplicate([1,2,3,4]))

File: week3.1/duplicate.js
function findDuplicate(nums) {
    // approach -1
    // =======================
    // const set = new Set()
    // for(let num of nums){
    //     if(set.has(num)){
    //         return true
    //     } else{
    //         set.add(num)
    //     }
    // }
    // return false
    // ==========================
    //approach -2
    // ==========================
    const map = new Map()
    for(let num of nums){
        if(map.has(num)){
            return true
        } else{
            map.set(num,1)
        }
    }
    return false
}

console.log(findDuplicate([1,2,3,4,5,6,7,8,9,10,1]))
console.log(findDuplicate([1,2,3,4,5,6,7,8,9,10]))